🐛 Fix validation error when `File` is declared after `Form` parameter

thomasleveil opened the pull request:
This PR updates the fix (6185a92) from @jsg921019 so that tests pass with the latest FastAPI code (0.110.0).
It also adds a test to cover the case with multiple files, and another to ensure the order of parameters in an endpoint definition does not affect the way FastAPI handles the request.

**Problem**
When declaring an endpoint which accepts a `File` but also a `Form` parameter, the `File` parameter must be declared first or the endpoint would respond with error 422.

**Context**
- See discussion #9116
- And probably also #4625

thomasleveil committed:
:white_check_mark: add tests for using `Form` and `File` in the same endpoint

YuriiMotov reviewed (changes requested):
Nothing critical (except import from typing).

I think it would be better to put all these related tests into one file and get rid of some of them (like `test_post_form_no_file` (and so on) for multiple files)

thomasleveil commented:
Thank you for the feed back. I pushed force with your suggested changes. Let me know if anything else can be improved.

YuriiMotov reviewed (approved):
The solution works.
Tests cover the issue. Tests fail before changes and pass after

cfreksen commented:
I am wondering if this PR has been forgotten?

I am not that familiar with the workflow in this repository, so excuse me if these things just takes time or if this has low priority and there have been more important things to spend your time on.

However, I was affected by #9116 which led me to this PR, and since it has some approvals and it hasn't been touched in 9 months, I thought it would be ok to "bump" it.

thomasleveil committed:
:bug: Fix handling uploaded file data correctly when not in the first part of a multipart/form-data request body

The HTTP specification for multipart/form-data, defined in RFC 7578, allows both form data and uploaded files in a single request. It does not require files to be before or after form fields.

As such, no specific care is to be given to the first field.

References:
- [RFC 7578](https://datatracker.ietf.org/doc/html/rfc7578)
- [FastAPI issue 9116](https://github.com/fastapi/fastapi/discussions/9116)

Close #9116

YuriiMotov commented:
> I don't get what you mean or refer to

I'm talking about `elif` block:
```python
        elif (
            is_bytes_sequence_field(field)
            and isinstance(field_info, params.File)
            and value_is_sequence(value)
        ):
```
It handles list of files and it was also modified. So, I think it should be covered by tests

YuriiMotov reviewed (approved):
LGTM!

tiangolo committed:
✅ Tweak test, add types

tiangolo committed:
Merge branch 'master' into fix-9116-form-and-file-order

tiangolo reviewed (approved):
Great, thank you! 🚀 🙌
